### PR TITLE
Switch from Kflow to use plain Postgres instead

### PR DESCRIPTION
**Pull request by onno-vos-dev**
- Implement buffer to get a sliding window of metrics,
  prefer dropping older metrics over newer ones.
- Add basic docker-compose for Postgres/Grafana testing/verification

### CODE REVIEW

File: src/system_monitor_pg.erl [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  epgsql:close(Conn).
+
+%%%_* Internal buffer functions ===============================================
+buffer_new() ->
+  [].
+
+buffer_add(Buffer, Element) ->
+  MaxBufferSize = application:get_env(?APP, max_buffer_size, 1000),
+  case buffer_size(Buffer) > MaxBufferSize of
+    true -> [Element | lists:reverse(tl(lists:reverse(Buffer)))];
```
Comment (mikpe) [inline src/system_monitor_pg.erl:104]: O(N^2) here. I think the buffer needs a bit more structure, but I'll need to think about the details for a while.

File: src/system_monitor_top.erl [hunk 2]
```diff
@@ -14,7 +14,7 @@
 %% limitations under the License.
 %%--------------------------------------------------------------------------------
 %%% @doc
-%%% Collect Erlang process statitics and push it to Kafka
+%%% Collect Erlang process statitics and push it to Postgres
```
Comment (mikpe) [inline src/system_monitor_top.erl:17]: nit1: s/statitics/statistics/g
nir2: here and elsewhere the comments have s/Kafka/Postgres/, but the destination is configurable with PG only being the built-in default; should it say "to the configured destination" or something like that?

File: src/system_monitor_pg.erl [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+-include_lib("system_monitor/include/system_monitor.hrl").
+-include_lib("hut/include/hut.hrl").
+
+-define(SERVER, system_monitor_pg).
+-define(FIVE_SECONDS, 5000).
+-define(ONE_HOUR, 60*60*1000).
+
+%%%_* API =================================================================
+start() ->
+  {ok, _Pid} = gen_server:start_link({local, ?SERVER}, ?MODULE, [], []),
```
Comment (mikpe) [inline src/system_monitor_pg.erl:25]: If I read this PR correctly, the system_monitor_pg gen_server is linked to the system_monitor gen_server, which feels like they're too closely coupled. Could system_monitor_pg instead add itself under system_monitor_sup?

File: src/system_monitor_pg.erl (hunk 1 above)
Comment (anha0825) [reply src/system_monitor_pg.erl:104]: We have talked about moving to queue.erl, but the length function will still be dominating. `{Size::non_neg_integer(), queue()}` could be a solution if we are concerned with the run time of this. Introducing such a data type would need some good testing.

File: src/system_monitor_pg.erl [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+   SS,
+   HS,
+   THS,
+   system_monitor:fmt_stack(CS),
+   GL].
+
+ts_to_timestamp(TS) ->
+  calendar:system_time_to_universal_time(TS, native).
+
+name_to_list(Term) ->
```
Comment (anha0825) [inline src/system_monitor_pg.erl:257]: This function doesn't work as expected for atoms.

File: docker/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+version: "3.6"
+networks:
```
Comment (anha0825) [inline docker/docker-compose.yml:2]: We should try to remove this network.

File: src/system_monitor_pg.erl (hunk 1 above)
Comment (onno-vos-dev) [reply src/system_monitor_pg.erl:104]: Implemented it as a `{Len, queue()}` now

File: src/system_monitor_pg.erl (hunk 3 above)
Comment (onno-vos-dev) [reply src/system_monitor_pg.erl:25]: Yes it could. Change upcoming :+1:

**Review by anha0825 — COMMENTED** [review]
system_monitor is published as a hex.pm package. We should decide how to handle that.

**Review by anha0825 — APPROVED** [review]

File: docker/db/30-db_schema.sh [hunk 6]
```diff
@@ -0,0 +1,126 @@
+#!/bin/bash
+set -euxo
+
+psql -v ON_ERROR_STOP=1 --username "system_monitor" --dbname "system_monitor" <<-EOSQL
```
[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply docker/db/30-db_schema.sh:4]: Exactly :) I'm, fine with that

File: rebar.config [hunk 7]
```diff
@@ -3,12 +3,12 @@
...
 {deps,
- [ {brod, "3.7.0"}
- , {supervisor3, "1.1.9"}
+ [ {supervisor3, "1.1.9"}
  , {hut, "1.3.0"}
+ , {epgsql, "4.5.0"}
  ]}.
 
-{dialyzer, [{warnings, [unknown]}]}.
+{dialyzer, [{warning, unknown}]}.
```
[the remark this answers is not in the document]
Comment (onno-vos-dev) [reply rebar.config]: Fixed

**Review by anha0825 — APPROVED** [review]

**Review by enidgjoleka — APPROVED** [review]

**Review by mikpe — APPROVED** [review]

**Comment by onno-vos-dev** [comment]
> system_monitor is published as a hex.pm package. We should decide how to handle that.

Yes we should

**Pull request merged**